### PR TITLE
Support custom CA on HTTP layer 

### PR DESCRIPTION
**Pull request by pebrc**
Fixes https://github.com/elastic/cloud-on-k8s/issues/4522

Allows users to add a full CA certificate in `ca.crt` and `ca.key` which will be used to sign the leaf certificate that ECK will use for the Elastic stack applications. 

Awkwardly the options we offer in `selfSignedCertificate` can then still be used to modify the certificate. Awkwardly, because that certificate is no longer necessarily self-signed, depending on the CA the user has provided. But I don't think we want to break the stable API we have for that. 

This also changes the naming convention for the transport layer CA user can specify for Elasticsearch to follow the new `ca.crt`, `ca.key` scheme. The old `tls.key` and `tls.crt` names are still supported for backwards compatibility.

Docs preview: https://cloud-on-k8s_4575.docs-preview.app.elstc.co/diff

### CODE REVIEW

File: pkg/controller/common/certificates/secret.go
Comment (david-kow) [inline pkg/controller/common/certificates/secret.go:123]: Took me a second to understand what `and` and `or` operate on :) Maybe something like below (+ change arguments order)?

```suggestion
		return fmt.Errorf("cannot specify %s or %s when %s is set in %s/%s",
```

File: pkg/controller/common/certificates/secret.go
Comment (david-kow) [inline pkg/controller/common/certificates/secret.go:128]: Change the func name (and callers) then? ie. `parseCustomCA`?

File: test/e2e/es/certs_test.go
Comment (david-kow) [inline test/e2e/es/certs_test.go:66]:
```suggestion
	// Now modify cluster to use custom HTTP CA but simulate a user error by populating the secret with
```

File: test/e2e/test/elasticsearch/builder.go
Comment (david-kow) [inline test/e2e/test/elasticsearch/builder.go:458]: Should we append instead? If one chains those only the last will be in effect which might be surprising.

File: pkg/controller/common/certificates/ca_secret.go
Comment (david-kow) [inline pkg/controller/common/certificates/ca_secret.go:15]: Should this function name (or at least godoc) indicate allowing legacy key naming? I'm not sure how one decides calling `ParseCustomCASecret` vs `parseCAFromSecret` (for example from `validateCustomCA`).

**Review by david-kow — APPROVED** [review]
LGTM, few nits here and there.

File: docs/orchestrating-elastic-stack-applications/accessing-elastic-services.asciidoc
Comment (pebrc) [reply docs/orchestrating-elastic-stack-applications/accessing-elastic-services.asciidoc:152]: I was honestly confused by the original example. Why would one want to specify the leaf certificate as the CA certificate? If the leaf certificate were also the CA at the same time we would not need to specify it. So I am thinking the original example was done by mistake. At the very least it would have required some additional explanation. So I made this change on purpose to reflect what I would think is a typical use case.

File: pkg/controller/common/certificates/ca_secret.go
Comment (pebrc) [reply pkg/controller/common/certificates/ca_secret.go:15]: Good point. I actually want to hide the `parseCAFromSecret` one as an internal function that just avoids duplication between the different parsing functions and meant `ParseCustomCASecret` as the actual API.

**Comment by pebrc** [comment]
The e2e test I added is flaky. I will need to fix that before merging this.

**Review by alaudazzi — APPROVED** [review]
LGTM

File: pkg/controller/common/certificates/secret.go
Comment (barkbay) [inline pkg/controller/common/certificates/secret.go:24]:
```suggestion
	// CAKeyFileName is used for CA certificate's private keys inside a secret.
```

File: pkg/controller/common/certificates/secret.go
Comment (barkbay) [inline pkg/controller/common/certificates/secret.go:105]: nit: at first I was thinking that a "full CA" is a CA with all the intermediate certificates. Should we name this function `HasPrivateKey` to make its purpose more explicit ?

File: docs/orchestrating-elastic-stack-applications/accessing-elastic-services.asciidoc
Comment (barkbay) [inline docs/orchestrating-elastic-stack-applications/accessing-elastic-services.asciidoc:155]: nit: I think I would add a comment to explain how `selfSignedCertificate` is used in the generated certificates.

**Review by barkbay — COMMENTED** [review]
Just left a few comments after a first read, I did not test it extensively.

File: pkg/apis/elasticsearch/v1/elasticsearch_types.go
Comment (barkbay) [inline pkg/apis/elasticsearch/v1/elasticsearch_types.go:103]: Should we keep these 2 lines ?  My understanding is that `tls.crt` and `tls.key` are still required if `ca.key` is not provided.

Comment (pebrc) [reply pkg/apis/elasticsearch/v1/elasticsearch_types.go:103]: This is for the transport layer. I replaced the `tls.crt` `tls.key` keys (but offer backwards compatibility for existing users). Your statement is true for the HTTP layer.

File: docs/orchestrating-elastic-stack-applications/accessing-elastic-services.asciidoc
Comment (pebrc) [reply docs/orchestrating-elastic-stack-applications/accessing-elastic-services.asciidoc:155]: This is actually one of the not so great aspects of this change. We will have a an attribute `selfSignedCertificate.subjectAltNames` that affects also certificates that are not actually self-signed.

File: pkg/apis/elasticsearch/v1/elasticsearch_types.go
Comment (barkbay) [reply pkg/apis/elasticsearch/v1/elasticsearch_types.go:103]: My bad, I missed the `TransportTLSOptions` in the diff header.

File: test/e2e/es/certs_test.go
Comment (barkbay) [inline test/e2e/es/certs_test.go:122]:
```suggestion
					Name: "Update custom CA secret with valid CA",
```

File: test/e2e/es/certs_test.go
Comment (barkbay) [inline test/e2e/es/certs_test.go:145]: How do we know from here that the operator has reconciled the cluster and the updated custom Secret is in use ?

Comment (pebrc) [reply test/e2e/es/certs_test.go:145]: I added a new check step called `CheckHTTPConnectivity` which should create a test request to each Pod. But you are right that there is a gap in the logic in that this step relies on the `*-es-certs-public` secret being reconciled and then uses the CA from that secret to make the connections.

Comment (pebrc) [reply test/e2e/es/certs_test.go:145]: I think could add an explicit step that verifies with the custom CA.

File: test/e2e/es/certs_test.go
Comment (barkbay) [inline test/e2e/es/certs_test.go:98]: I would include a namespaced name comparison to check that the event has not been generated by a previous run (it happened on my dev environment), or a similar test that we may add in the future.
```suggestion
								evt.Reason == events.EventReconciliationError &&
								evt.InvolvedObject.Namespace == b.Elasticsearch.Namespace &&
								evt.InvolvedObject.Name == b.Elasticsearch.Name &&
```

File: test/e2e/es/certs_test.go
Comment (barkbay) [reply test/e2e/es/certs_test.go:145]: Also if I slow down the operator and delay the reconciliation I can see this kind of error:

```
    cluster_uuid.go:57: 
        	Error Trace:	cluster_uuid.go:57
        	Error:      	Received unexpected error:
        	            	Get "https://test-custom-http-ca-2fxk-es-http.e2e-mercury.svc:9200/": x509: certificate signed by unknown authority
        	Test:       	TestCustomHTTPCA/Cluster_UUID_should_have_been_preserved#01
```

Not sure if it is a realistic experiment though...

Comment (pebrc) [reply test/e2e/es/certs_test.go:145]: Slowing down means you also delayed the certificate secret reconciliation to run into the condition I [mentioned](https://github.com/elastic/cloud-on-k8s/pull/4575#discussion_r659671813)?

Comment (barkbay) [reply test/e2e/es/certs_test.go:145]: Yes, I think `CheckHTTPConnectivity` was run before the certificate was reconciled:
 
image

You're right, it is just another manifestation of the same condition. Sorry, I'm just trying to understand if this test could become a flaky one once deployed 🙃

Comment (pebrc) [reply test/e2e/es/certs_test.go:145]: No need to be sorry. Very grateful for your thorough review!

Comment (barkbay) [reply test/e2e/es/certs_test.go]: I just checked out last commit ([5e29d98](https://github.com/elastic/cloud-on-k8s/pull/4575/commits/5e29d98ac5c11745d9ccb5649a9e08f54baa1ac6)), it looks good but I think we need a similar step to validate that at step `reconfigure back to self-signed operator provided CA` the custom CA is no more in use.

File: test/e2e/es/certs_test.go
Comment (barkbay) [inline test/e2e/es/certs_test.go:165]: Any reason for having this check as a `PostMutationStep` ? I think we should expect the CA to be available right after the update, something along those lines:

```golang
PreMutationSteps: func(k *test.K8sClient) test.StepList {
    return test.StepList{
        {
            Name: "Update custom CA secret",
            Test: test.Eventually(func() error {
                var err error
                customCA, err = certificates.NewSelfSignedCA(certificates.CABuilderOptions{
                    Subject: pkix.Name{
                        CommonName:         "eck-e2e-test-custom-ca",
                        OrganizationalUnit: []string{"eck-e2e"},
                    },
                })
                if err != nil {
                    return err
                }

                caSecret := mkCertSecret(
                    certificates.EncodePEMCert(customCA.Cert.Raw),
                    certificates.EncodePEMPrivateKey(*customCA.PrivateKey),
                )
                _, err = reconciler.ReconcileSecret(k.Client, caSecret, nil)
                return err
            }),
        },
        {
            // Updated CA should be exposed almost immediately 
            Name: "Verify that the custom CA is in use",
            Test: test.Eventually(func() error {
                return elasticsearch.CheckHTTPConnectivityWithCA(initialCluster.Elasticsearch, k, []*x509.Certificate{customCA.Cert})
            }),
        },
    }
},
```

File: pkg/controller/common/certificates/secret_test.go
Comment (barkbay) [inline pkg/controller/common/certificates/secret_test.go:195]:
```suggestion
			name: "With CA private key",
```

**Review by barkbay — APPROVED** [review]
LGTM 👍

File: test/e2e/es/certs_test.go
Comment (pebrc) [reply test/e2e/es/certs_test.go:165]: Good point!

Comment (pebrc) [reply test/e2e/es/certs_test.go:165]: I guess the only reason would be that it allows you to switch the sequence of the builders without breaking the test.

**Comment by pebrc** [comment]
run full pr build

**Comment by pebrc** [comment]
@barkbay I reworked the e2e test to avoid spurious test failures. It is much simpler now and should be stable  🤞 .  If you could take another quick look that would be great 🙇

File: test/e2e/es/certs_test.go
Comment (barkbay) [inline test/e2e/es/certs_test.go:135]:
```suggestion
					return errors.New("still using custom CA cert")
```

File: test/e2e/test/elasticsearch/checks_http.go
Comment (barkbay) [inline test/e2e/test/elasticsearch/checks_http.go:44]: While re-reading this check I was wondering if we could "miss" some Pods, either because some of them are being recreated or because they are not ready.
I guess that if we do not want to make any progress until all the expected Pods are exposing the desired CA the algorithm would be:

1. Get the actual nodes, if at least one is not available/ready return an error
2. Get all the expected nodes, if the list does not match the actual one return an error
3. For each node check the connectivity using the desired CA

This situation is very unlikely to happen though, so I'm not sure if it's worth to improve this check.

File: test/e2e/es/certs_test.go
Comment (barkbay) [inline test/e2e/es/certs_test.go:59]: cosmetic nit: the `steps` variable can be avoided

```suggestion
	test.StepList{}.
		WithSteps(initialCluster.InitTestSteps(k)).
		WithSteps(initialCluster.CreationTestSteps(k)).
		WithSteps(test.CheckTestSteps(initialCluster, k)).
		WithStep(test.Step{
			Name: "Create an invalid CA secret",
			Test: test.Eventually(func() error {
				bogusSecret := mkCertSecret([]byte("garbage"), []byte("more garbage"))
				_, err := reconciler.ReconcileSecret(k.Client, bogusSecret, nil)
				return err
			}),
		}).
		WithSteps(withCustomCA.UpgradeTestSteps(k)).
		WithStep(
			test.Step{
				Name: "Invalid CA secret should create events",
				Test: test.Eventually(func() error {
					eventList, err := k.GetEvents(test.EventListOptions(esNamespace, initialCluster.Elasticsearch.Name)...)
					if err != nil {
						return err
					}
					for _, evt := range eventList {
						if evt.Type == corev1.EventTypeWarning &&
							evt.Reason == events.EventReconciliationError &&
							evt.InvolvedObject.Namespace == withCustomCA.Elasticsearch.Namespace &&
							evt.InvolvedObject.Name == withCustomCA.Elasticsearch.Name &&
							strings.Contains(evt.Message, "can't parse") {
							return nil
						}
					}
					return fmt.Errorf("expected validation event but could not observe it")
				})}).
		WithStep(test.Step{
			Name: "Update custom CA secret",
			Test: test.Eventually(func() error {
				var err error
				customCA, err = certificates.NewSelfSignedCA(certificates.CABuilderOptions{
					Subject: pkix.Name{
						CommonName:         "eck-e2e-test-custom-ca",
						OrganizationalUnit: []string{"eck-e2e"},
					},
				})
				if err != nil {
					return err
				}

				caSecret := mkCertSecret(
					certificates.EncodePEMCert(customCA.Cert.Raw),
					certificates.EncodePEMPrivateKey(*customCA.PrivateKey),
				)
				_, err = reconciler.ReconcileSecret(k.Client, caSecret, nil)
				return err
			}),
		}).
		WithStep(test.Step{
			Name: "Verify that the custom CA is in use",
			Test: test.Eventually(func() error {
				return elasticsearch.CheckHTTPConnectivityWithCA(initialCluster.Elasticsearch, k, []*x509.Certificate{customCA.Cert})
			}),
		}).
		// "upgrade" back to the initial cluster without custom CA
		WithSteps(initialCluster.UpgradeTestSteps(k)).
		WithSteps(test.StepList{
			{
				Name: "Ensure built-in CA is used after removing custom CA",
				Test: test.Eventually(func() error {
					ca, err := k.GetCA(initialCluster.Elasticsearch.Namespace, initialCluster.Elasticsearch.Name, certificates.HTTPCAType)
					if err != nil {
						return err
					}
					if customCA != nil && customCA.Cert.Equal(ca.Cert) {
						return errors.New("still using custom CA cert")
					}
					return elasticsearch.CheckHTTPConnectivityWithCA(initialCluster.Elasticsearch, k, []*x509.Certificate{ca.Cert})
				}),
			},
			{
				Name: "Clean up custom CA secret",
				Test: func(t *testing.T) {
					// let's clean up the CA secret here.
					toDelete := mkCertSecret(nil, nil)
					_ = k.Client.Delete(context.Background(), &toDelete)
				},
			},
		}).WithSteps(initialCluster.DeletionTestSteps(k)).RunSequential(t)
```

**Review by barkbay — APPROVED** [review]
lgtm, checking the CA right after a related change should greatly improve the stability 👍

File: test/e2e/test/elasticsearch/checks_http.go
Comment (pebrc) [reply test/e2e/test/elasticsearch/checks_http.go:44]: I think you are right that is unlikely. All upgrades in this test are not leading to actual Pod rotations and we check the readiness of the expected Pods during the initial creation. So I am tempted to leave it as is for now.

**Pull request merged**
